internal/step/plugin/provider.go

```diff
@@ -1100,78 +1308,125 @@ func (r *runningStep) startStage(container deployer.Plugin) error {
 			r.state = step.RunningStepStateRunning
 			r.lock.Unlock()
 		case <-r.ctx.Done():
-			return fmt.Errorf("step closed while waiting for run configuration")
+			r.logger.Debugf("step closed while waiting for run configuration")
+			return true, 0, nil
 		}
 	}
 
 	inputSchema, err := r.atpClient.ReadSchema()
 	if err != nil {
-		return err
+		return false, 0, err
 	}
 	steps := inputSchema.Steps()
 	stepSchema, ok := steps[r.pluginStepID]
 	if !ok {
-		return fmt.Errorf("error in run step %s: schema mismatch between local and remote deployed plugin, no stepSchema named %s found in remote", r.runID, r.pluginStepID)
+		return false, 0, fmt.Errorf("error in run step %s: schema mismatch between local and remote deployed plugin, no stepSchema named %s found in remote", r.runID, r.pluginStepID)
 	}
 	// Re-verify input. This should have also been done earlier.
-	if _, err := stepSchema.Input().Unserialize(runInput); err != nil {
-		return fmt.Errorf("schema mismatch between local and remote deployed plugin in step %s/%s, unserializing input failed (%w)", r.runID, r.pluginStepID, err)
+	if _, err := stepSchema.Input().Unserialize(runInput.stepInputData); err != nil {
+		return false, 0, fmt.Errorf("schema mismatch between local and remote deployed plugin in step %s/%s, unserializing input failed (%w)", r.runID, r.pluginStepID, err)
 	}
 
+	r.wg.Add(1)
+
 	// Runs the ATP client in a goroutine in order to wait for it.
-	// On context done, the deployer has 30 seconds before it will error out.
+	// On context done, the deployer has limited time before it will error out.
 	go func() {
+		defer r.wg.Done()
 		result := r.atpClient.Execute(
-			schema.Input{RunID: r.runID, ID: r.pluginStepID, InputData: runInput},
+			schema.Input{RunID: r.runID, ID: r.pluginStepID, InputData: runInput.stepInputData},
 			r.signalToStep,
 			r.signalFromStep,
 		)
+		r.lock.Lock()
+		// The sender should be the one to close the signal send channel
+		channel := r.signalToStep
+		r.signalToStep = nil
+		close(channel)
+		r.lock.Unlock()
 		r.executionChannel <- result
 		if err = r.atpClient.Close(); err != nil {
 			r.logger.Warningf("Error while closing ATP client: %s", err)
 		}
 	}()
-	return nil
+	return false, runInput.forceCloseTimeoutMS, nil
 }
 
-func (r *runningStep) runStage() error {
+func (r *runningStep) runStage(forCloseWaitMS int64) error {
```

Comment:
"forClose" sounds like a bad mortgage, and the similarity to `forceCloseTimeoutMS` makes this at first glance look like a typo. Maybe `waitForCloseMS`?